src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -2426,21 +2426,57 @@ public State getState() {
         return threadState();
     }
 
+    /**
+     * Returns the translation from a J9VMThread state to a Thread::State.
+     *
+     * @return this thread's state.
+     *
+     * @see State
+     */
+    private State translateJ9VMThreadStateToThreadState(int status) {
+        switch (status) {
+            case 1: // J9VMTHREAD_STATE_RUNNING
+                return State.RUNNABLE;
+            case 2: // J9VMTHREAD_STATE_BLOCKED
+                return State.BLOCKED;
+            case 4: // J9VMTHREAD_STATE_WAITING
+            case 0x80: // J9VMTHREAD_STATE_PARKED
+                return State.WAITING;
+            case 8: // J9VMTHREAD_STATE_SLEEPING
+            case 64: // J9VMTHREAD_STATE_WAITING_TIMED
+            case 0x100: // J9VMTHREAD_STATE_PARKED_TIMED
+                return State.TIMED_WAITING;
+            case 32: // J9VMTHREAD_STATE_DEAD
+                return State.TERMINATED;
+            default:
+                synchronized (interruptLock) {
+                    if (eetop == NO_REF) {
+                        return State.TERMINATED;
+                    }
+                    return State.values()[getStateImpl(eetop)];
+                }
+        }
+    }
+
     /**
      * Returns the state of this thread.
      * This method can be used instead of getState as getState is not final and
      * so can be overridden to run arbitrary code.
      */
     State threadState() {
-        synchronized (interruptLock) {
+        if (started) {
             if (eetop == NO_REF) {
-                if (isDead()) {
+                return State.TERMINATED;
+            }
+            if (holder == null) {
+                if (eetop == NO_REF) {
                     return State.TERMINATED;
                 }
-                return State.NEW;
+                return State.values()[getStateImpl(eetop)];
             }
-            return State.values()[getStateImpl(eetop)];
+            return translateJ9VMThreadStateToThreadState(holder.threadStatus);
```

Comment:
Please explain how this can work: I don't see anything in the VM that uses 
`J9VMJAVALANGTHREADFIELDHOLDER(_SET)?_THREADSTATUS`.

Reply:
yes, I've update the comment to depend, sorry for the confusion.